src/test/unittest/valgrind.py

```diff
@@ -0,0 +1,200 @@
+#       from this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+# A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+# OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+# SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+# LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+# DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+# THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+#
+"""Valgrind handling tools"""
+
+import sys
+import re
+import subprocess as sp
+from enum import Enum, unique
+from os import path
+
+from helpers import ROOTDIR, VMMALLOC
+
+
+DISABLE = -1
+ENABLE = 1
+AUTO = 0
+
+_IGNORED = (
+    "WARNING: Serious error when reading debug info",
+    "When reading debug info from ",
+    "Ignoring non-Dwarf2/3/4 block in .debug_info",
+    "Last block truncated in .debug_info; ignoring",
+    "parse_CU_Header: is neither DWARF2 nor DWARF3 nor DWARF4",
+    "brk segment overflow",
+    "see section Limitations in user manual",
+    "Warning: set address range perms: large range",
+    "further instances of this message will not be shown",
+    "get_Form_contents: DW_FORM_GNU_strp_alt used, but no alternate .debug_str"
+)
+
+
+@unique
+class _Tool(Enum):
+    MEMCHECK = 1
+    PMEMCHECK = 2
+    HELGRIND = 3
+    DRD = 4
+    NONE = 5
+
+
+TOOLS = tuple(t for t in _Tool if t != _Tool.NONE)
+
+
+def enabled_tool(test):
+    """Get valgrind tool enabled by test"""
+    enabled = [t for t in TOOLS if getattr(test, t.name.lower()) == ENABLE]
+    if len(enabled) > 1:
+        raise ValueError('test "{}" enables more than one valgrind tool'
+                         .format(test))
+    elif len(enabled) == 1:
+        return enabled[0]
+    else:
+        return None
+
+
+def disabled_tools(test):
+    """Get valgrind tools disabled by test"""
+    disabled = [t for t in _Tool if getattr(test, t.name.lower()) == DISABLE]
+    return disabled
+
+
+class Valgrind:
+    def __init__(self, tool, cwd, testnum, memcheck_check_leaks=True):
+        self.tool = _Tool.NONE if tool is None else tool
+        self.cwd = cwd
+
+        log_file_name = '{}{}.log'.format(self.tool.name.lower(), testnum)
+        self.log_file = path.join(cwd, log_file_name)
+        if sys.platform != 'win32':
+            self.valgrind_exe = self._get_valgrind_exe()
+        else:
+            self.valgrind_exe = ''
+        self.tool_opts = ''
+
+        self.add_suppression('ld.supp')
+        self.add_suppression('memcheck-libunwind.supp')
+        self.add_suppression('memcheck-ndctl.supp')
```

Comment:
please unify suppression setup

the way unittest.sh does this seems to be a bug